### ISSUE TITLE
Support of non-UTF-8 psk identity

### ISSUE THREAD

**Issue by AndreasHuber-CH** [eclipse/californium#902]
We have a DTLS client that does not respect the RFC by using PSK identities that are not valid UTF-8 byte sequences.
In that case the code "identity = new String(this.identityEncoded, UTF_8)" in PSKClientKeyExchange corrupts the key identity.

We would like a way to support those non-standard clients. For example by being able to specify the PSK identity charset encoding.
Do you think this would be possible?

**Comment by boaks** [comment]
Possible, yes. 
But are you sure, that you want to start with such non-compliant clients instead of fixing them before starting?

**Comment by boaks** [comment]
To prevent misunderstandings: 

If it's required and the clients can not be changed,  the best solution would be for me to change the interface `PskStore` by replacing the "String identity" with "byte[] identity" and create a abstract class to feed the current implementations with Strings. That breaks the API. So I will only do that, if it's really required for you. If it is a non-compliant NB-IoT module, I guess you will not be the last to ask for it.

If you can change the client, that would be my preferred solution.

**Comment by AndreasHuber-CH** [comment]
You describe our issue pretty well. We can motivate clients to become compliant, but we'll have to continue to support some non-standard clients at least for some time.
We were using the DTLS implementation from Bouncy Castle which treats the identity as a "byte[]" and only now when trying to move to Scandium we realized that we actually have non-standard clients.

Changing the PskStore interface would be quite a proper solution (I had a less intrusive but more hacky solution in mind, like configuring a static Charset through a static setter directly on the PSKClientKeyExchange class).
But if you change PskStore. you would still keep the String identity in all the other places I suppose? (Like in the PreSharedKeyIdentity for example, and a lot of log messages).
This would be fine for us.

If you prefer to keep backwards compatibility you could always add a new interface BinaryPskStore. In this case the DtlsConnectorConfig could accept either a PskStore or a BinaryPskStore and if it receives a PskStore it could wrap it into a class that implements BinaryPskStore and delegates to the given PskStore.
(But this adds a bit more more boiler plate to Scandium, versus an abstract PskStore class like you proposed. In java 8 you could have added a default method to the PskStore interface, but I think californium must stay java 1.7 compatible?)

**Comment by boaks** [comment]
> But if you change PskStore. you would still keep the String identity in all the other places I suppose? 

I think, we should replace it every where, and ensure, that logging will not throw an exception. 
I would currently consider to use a subclass of `org.eclipse.californium.elements.util.Bytes` and override the `toString()`. I will prepare a PR for it. I think it will change some more files, so I will do that after the open PR are merged or withdrawn.

 > but I think californium must stay java 1.7 compatible?

That's the 2.0.x promise ...

**Comment by AndreasHuber-CH** [comment]
Ok, yes adding a PskIdentity class that is properly supported throughout scandium is even better.
A good toString() implementation is maybe more difficult. For backwards compatibility I think you'd want continue to log "new String(this.identityEncoded, UTF_8)". This does not throw an exception for non-UTF-8 byte input but rather logs an ugly string (strange characters, and invalid octets are replaced by some symbol).
Maybe you could log the PskIdentity as a hex string prefixed by 0x if the byte array is not valid UTF-8? But then if a 'binary psk identity' is valid UTF-8 by chance, the logged value might still contain strange some characters.

Thanks a lot to plan this PR. I understand well that it might not be the most important PR to fix.

**Comment by boaks** [comment]
The identities in the stores may be created with a `byte[]` and the `String` and I plan to use them, where ever it is possible. So it's left to the north-bound-system to provide the best string for the expected `byte[]` identities.

There are not too many place, when the `byte[]` received before the lookup is triggered, where the `String` is used. It should be acceptable to have there a "similar" String in the logs.

**Comment by sbernard31** [comment]
My 2 cents, I think default pskstore should not accept invalid psk identity.
But user could use their own stores which could support invalid psk identity (for reason exposed above).

**Comment by boaks** [comment]
Please checkout the `psk_bytes_id` branch.
The `PskPublicInformation` enables to compare with byte arrays and use a "proper" string for the `PreSharedKeyIdentity` after successful  matching the bytes.
It changes a "little more" and I didn't update javadoc etc. so it's for now just a wip branch.
Therefore, please have a first look and try to check, if that helps you. If it works, I would go to polish it and prepare a PR.

**Comment by boaks** [comment]
New introduced `PskPublicInformation`:
https://github.com/eclipse/californium/blob/psk_bytes_id/scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/PskPublicInformation.java

Normalize Identity on secret key lookup:
https://github.com/eclipse/californium/blob/psk_bytes_id/scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/pskstore/InMemoryPskStore.java#L101

Add normalized Identity to bytes psk store:
https://github.com/eclipse/californium/blob/psk_bytes_id/scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/pskstore/InMemoryPskStore.java#L124

**Comment by AndreasHuber-CH** [comment]
I tested the `psk_bytes_id` branch. The feature works. I could validate that a non-compliant client can successfully connect to the DTLS server. Thanks a lot !

I have just a comment about the `PskPublicInformation.normalize()` function:
If I call this function inside my `BytesPskStore.getKey()` implementation I can provide a custom String identity for the non-UTF-8 compliant identities which is then be passed to the `PreSharedKeyIdentity` and used elsewhere.
But among the non-compliant clients, there might be some that might use a valid UTF-8 byte sequence by chance, but one that prints as `?????????????????` in the log (I tested with the psk identity ᚠᛇᚻ᛫ᛒᛦᚦ᛫ᚠᚱᚩᚠᚢᚱ᛫ᚠᛁ that prints as ????????????????? in my eclipse console).

For uniformity it might be nice if `PskPublicInformation.normalize()` can be called independently whether `PskPublicInformation.isUtf8Compliant()` returns true or false, similar to the `PskPublicInformation(String identity, byte[] identityBytes)` constructor that accepts passing a custom String identity also for a UTF-8 compliant identityBytes value. E.g. like:
```java
    public void normalize(String publicInfo) {
        this.publicInfo = publicInfo;
        utf8Compliant = Arrays.equals(getBytes(), publicInfo.getBytes(UTF_8));
    }
```

This way you should rather check `psk.identity.isUtf8Compliant()` instead of `identity.isUtf8Compliant()` when calling `normalize()` in the `InMemoryPskStore` to align the two identities:
```java
		if (keyMap != null) {
			Psk psk = keyMap.get(identity);
			if (psk != null) {
				if (!psk.identity.isUtf8Compliant()) {
					identity.normalize(psk.identity.getPublicInfoAsString());
				}
				return psk.getKey();
			}
		}
```

My proposition changes slightly the meaning of the `utf8Compliant` flag and makes it mutable. But would allow transforming all client PSK identies into "hex string" and not only the non-UTF-8 compliant ones for example.
What do you think?

**Comment by boaks** [comment]
Thanks for your feedback!
The proposed change makes sense :-).
I will adapt the wip and start to polish it next week.

**Comment by AndreasHuber-CH** [comment]
Great! Thanks a lot.

**Comment by boaks** [comment]
See PR #918

**Comment by boaks** [comment]
Fixed with PR #918
